src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java

```diff
@@ -0,0 +1,336 @@
+            }
+
+            String chainTaskId = task.getChainTaskId();
+
+            boolean blockNumberAvailable = task.getInitializationBlockNumber() != 0
+                    && task.getInitializationBlockNumber() <= blockNumber;
+            boolean hasWorkerAlreadyParticipated = replicatesService.hasWorkerAlreadyParticipated(
+                    chainTaskId, walletAddress);
+            boolean moreReplicatesNeeded = replicatesService.moreReplicatesNeeded(chainTaskId,
+                    task.getNumWorkersNeeded(), task.getMaxExecutionTime());
+
+            if (blockNumberAvailable && !hasWorkerAlreadyParticipated && moreReplicatesNeeded) {
+                replicatesService.addNewReplicate(chainTaskId, walletAddress);
+                workerService.addChainTaskIdToWorker(chainTaskId, walletAddress);
+                return replicatesService.getReplicate(chainTaskId, walletAddress);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    public List<InterruptedReplicateModel> getInterruptedReplicates(long blockNumber, String walletAddress) {
+
+        List<String> chainTaskIdList = workerService.getChainTaskIds(walletAddress);
+        List<Task> tasksWithWorkerParticipation = taskService.getTasksByChainTaskIds(chainTaskIdList);
+        List<InterruptedReplicateModel> interruptedReplicates = new ArrayList<>();
+
+        for (Task task : tasksWithWorkerParticipation) {
+
+            Optional<Replicate> oReplicate = replicatesService.getReplicate(
+                    task.getChainTaskId(), walletAddress);
+            if (!oReplicate.isPresent()) continue;
+
+            Replicate replicate = oReplicate.get();
+
+            boolean isRecoverable = replicate.isRecoverable();
+            if (!isRecoverable) continue;
+
+            RecoveryAction recoveryAction = getAppropriateRecoveryAction(task, replicate, blockNumber);
+            if (recoveryAction == null) continue;
+
+            // generate contribution authorization
+            Optional<ContributionAuthorization> authorization = signatureService.createAuthorization(
+                    walletAddress, task.getChainTaskId(), TeeUtils.isTrustedExecutionTag(task.getTag()));
+
+            InterruptedReplicateModel interruptedReplicate = InterruptedReplicateModel.builder()
+                    .contributionAuthorization(authorization.get())
+                    .recoveryAction(recoveryAction)
+                    .build();
+
+            // change replicate status
+            replicatesService.updateReplicateStatus(task.getChainTaskId(), walletAddress,
+                    ReplicateStatus.RECOVERING, ReplicateStatusModifier.POOL_MANAGER);
+
+            interruptedReplicates.add(interruptedReplicate);
+        }
+
+        return interruptedReplicates;
+    }
+
+    public RecoveryAction getAppropriateRecoveryAction(Task task, Replicate replicate, long blockNumber) {
+        ChainReceipt chainReceipt = new ChainReceipt(blockNumber, "");
```

Comment:
This one is not needed.

Reply:
It is actually, because in the case where ReplicateStatus is CONTRIBUTING and the worker contributes onchain without notifying the core should update the status to CONTRIBUTED and tell the worker to WAIT or REVEAL. So when updating the status to CONTRIBUTED we check if it's correct onchain, thus, we need the chainReceipt with the blocknumber.

Reply:
Yes, I can see it is used in methods _recoverReplicateInContributionPhase_ and _recoverReplicateInRevealPhase_ but we don't modify anything to this ChainReceipt. We better pass the _blocknumber_ to those methods and create a _new ChainReceipt(blockNumber, "")_ at the lowest level inside those two methods.

Reply:
true yeah, I see your point